src/contrasts.jl

```diff
@@ -0,0 +1,215 @@
+            ArgumentError(
+                "only one level found: $(c_levels[1]) (need at least two to " *
+                "compute contrasts).",
+            ),
+        )
+    end
+
+    # find index of base level. use contrasts.base, then default (1).
+    base_level = baselevel(contrasts)
+    baseind = base_level === nothing ? 1 : findfirst(isequal(base_level), c_levels)
+    if baseind === nothing
+        throw(ArgumentError("base level $(base_level) not found in levels " * "$c_levels."))
+    end
+
+    tnames = termnames(contrasts, c_levels, baseind)
+
+    mat = contrasts_matrix(contrasts, baseind, n)
+
+    return ContrastsMatrix(mat, tnames, c_levels, contrasts)
+end
+
+function ContrastsMatrix(c::Type{<:AbstractContrasts}, levels::AbstractVector)
+    return throw(
+        ArgumentError("contrast types must be instantiated (use $c() instead of $c)")
+    )
+end
+
+# given an existing ContrastsMatrix, check that all passed levels are present
+# in the contrasts. Note that this behavior is different from the
+# ContrastsMatrix constructor, which requires that the levels be exactly the same.
+function ContrastsMatrix(c::ContrastsMatrix, levels::AbstractVector)
+    if !isempty(setdiff(levels, c.levels))
+        throw(
+            ArgumentError(
+                "there are levels in data that are not in ContrastsMatrix: " *
+                "$(setdiff(levels, c.levels))" *
+                "\n  Data levels: $(levels)" *
+                "\n  Contrast levels: $(c.levels)",
+            ),
+        )
+    end
+    return c
+end
+
+function termnames(C::AbstractContrasts, levels::AbstractVector, baseind::Integer)
+    not_base = [1:(baseind - 1); (baseind + 1):length(levels)]
+    return levels[not_base]
+end
+
+function Base.getindex(contrasts::ContrastsMatrix{C,T}, rowinds, colinds) where {C,T}
+    return getindex(contrasts.matrix, getindex.(Ref(contrasts.invindex), rowinds), colinds)
+end
+
+"""
+    DummyCoding([base[, levels]])
+    DummyCoding(; base=nothing, levels=nothing)
+
+Dummy coding generates one indicator column (1 or 0) for each non-base level.
+
+If `levels` are omitted or `nothing`, they are determined from the data
+by calling the `levels` function on the data when constructing `ContrastsMatrix`.
+If `base` is omitted or `nothing`, the first level is used as the base.
+
+Columns have non-zero mean and are collinear with an intercept column (and
+lower-order columns for interactions) but are orthogonal to each other. In a
+regression model, dummy coding leads to an intercept that is the mean of the
+dependent variable for base level.
+
+Also known as "treatment coding" or "one-hot encoding".
+"""
+mutable struct DummyCoding <: AbstractContrasts
+    base::Any
+    levels::Union{AbstractVector,Nothing}
+end
+## constructor with optional keyword arguments, defaulting to nothing
+function DummyCoding(; base=nothing, levels::Union{AbstractVector,Nothing}=nothing)
+    return DummyCoding(base, levels)
+end
```

Comment:
Is this the only constructor and do you want people not to use any other constructor? Then, you can use the inner constructor and save the comment (https://docs.julialang.org/en/v1/manual/constructors/#man-inner-constructor-methods).

Reply:
Yes it is the only constructor, I will change it

Reply:
Done in ce0ffa3